app/scripts/lib/transaction/handlers.ts

```diff
@@ -0,0 +1,227 @@
+import type { Provider } from '@metamask/network-controller';
+import { FetchGasFeeEstimateOptions } from '@metamask/gas-fee-controller';
+import { safelyExecute } from '@metamask/controller-utils';
+
+import {
+  TokenStandard,
+  TransactionMeta,
+} from '../../../../shared/constants/transaction';
+import {
+  MetaMetricsEventFragment,
+  MetaMetricsPageObject,
+  MetaMetricsReferrerObject,
+} from '../../../../shared/constants/metametrics';
+import { type SnapAndHardwareMessenger } from '../snap-keyring/metrics';
+
+import * as metrics from './metrics';
+
+export type TransactionEventRequest = {
+  createEventFragment: (
+    options: MetaMetricsEventFragment,
+  ) => MetaMetricsEventFragment;
+  finalizeEventFragment: (
+    fragmentId: string,
+    options?: {
+      abandoned?: boolean;
+      page?: MetaMetricsPageObject;
+      referrer?: MetaMetricsReferrerObject;
+    },
+  ) => void;
+  getEventFragmentById: (fragmentId: string) => MetaMetricsEventFragment;
+  updateEventFragment: (
+    fragmentId: string,
+    payload: Partial<MetaMetricsEventFragment>,
+  ) => void;
+  getAccountType: (
+    address: string,
+  ) => Promise<'hardware' | 'imported' | 'MetaMask'>;
+  getDeviceModel: (
+    address: string,
+  ) => Promise<'ledger' | 'lattice' | 'N/A' | string>;
+  // According to the type GasFeeState returned from getEIP1559GasFeeEstimates
+  // doesn't include some properties used in buildEventFragmentProperties,
+  // hence returning any here to avoid type errors.
+  getEIP1559GasFeeEstimates(options?: FetchGasFeeEstimateOptions): Promise<any>;
+  getParticipateInMetrics: () => boolean;
+  getSelectedAddress: () => string;
+  getTokenStandardAndDetails: () => {
+    decimals?: string;
+    balance?: string;
+    symbol?: string;
+    standard?: TokenStandard;
+  };
+  getTransaction: (transactionId: string) => TransactionMeta;
+  provider: Provider;
+  snapAndHardwareMessenger: SnapAndHardwareMessenger;
+  trackEvent: (payload: any) => void;
+};
+
+export type TransactionEventPayload = {
+  transactionMeta: TransactionMeta;
+  actionId?: string;
+  error?: string;
+  approvalTransactionMeta?: TransactionMeta;
+};
+
+/**
+ * This function is called when a transaction is added to the controller.
+ *
+ * @param TransactionEventRequest - Contains controller actions needed to create/update/finalize event fragments
+ * @param transactionEventPayload - The event payload
+ * @param transactionEventPayload.transactionMeta - The transaction meta object
+ */
+export const handleTransactionAdded = async (
```

Comment:
I really appreciate the goal of splitting the code and modularising more, however in the case of event listeners, I think a big part of their power is decoupling the code that wishes to add a listener.

I worry that by forcing all the new logic to go into this separate file, we're just adding another level of indirection to determine who is interested in the events, and creating another file that could end up coupled to many different classes. We're also breaking the concept of listeners a little by creating mega listeners with many actions (in future) rather than many separate listeners with their own logic.

If we leave the listeners in `MetaMaskController`, it'll likely be easier to determine who is listening to the events, plus hopefully more explicit as they can live alongside the construction of the `TransactionController`.

As a compromise, we could at least create a private method in the `MetaMaskController` to add all the `TransactionController` listeners?

Reply:
Rolled back in https://github.com/MetaMask/metamask-extension/pull/21456/commits/86dfac2fc9acf571f3240e2ea59948f8ed6b4a1f